### PR TITLE
[feat] Support ProFit: Extend DFT with Probability Threshold-based Token Filtering

### PR DESCRIPTION
**Pull request by maybefunctionname**
# PR type
- [ ] Bug Fix
- [x] New Feature
- [ ] Document Updates
- [ ] More Models or Datasets Support

#### Overview
This PR extends the existing DFT (Dynamic Fine-Tuning) implementation to support the hard gating mechanism proposed in [ProFit](https://arxiv.org/pdf/2601.09195), which directly masks low-value tokens via probability thresholds to improve model generalization on reasoning tasks.

#### Key Improvements
**DFT vs ProFit - Core Differences:**
- **DFT (Soft Gating)**: Continuously reweights loss by token probability `p` (`loss *= p`), all tokens still contribute to gradients
- **ProFit (Hard Gating)**: Sets threshold `τ` to directly mask tokens with `p < τ` (`loss *= mask`), retaining only high-probability core tokens

**Advantages of ProFit:**
1. **Prevents Surface-level Overfitting**: Low-probability tokens often represent non-essential expressions (e.g., filler words, redundant modifiers). Forcing alignment with them distracts the model from core logic
2. **Stronger Generalization**: Experiments show ProFit achieves 3.0%-10.9% average improvement over standard SFT on reasoning benchmarks like GPQA-Diamond and MATH-500
3. **Theoretical Guarantee**: The paper proves low-probability tokens induce larger gradients that can overshadow key semantic signals

#### Implementation Details
To maintain code simplicity, this implementation reuses the `enable_dft_loss` parameter and switches modes via the environment variable `HARD_GATING_PROBABILITY_THRESHOLD`:
- **Variable not set**: Uses original DFT soft gating
- **Set to float (e.g., `0.1`)**: Enables ProFit hard gating, masking tokens below threshold

```python
# Core modification
if hard_gating_probability_threshold := os.getenv("HARD_GATING_PROBABILITY_THRESHOLD"):
    mask = (target_probs > float(hard_gating_probability_threshold)).float()
    target_probs = mask * (labels != -100).float()  # Hard gating
```

#### Usage Example
```bash
# Enable ProFit (threshold 0.1)
HARD_GATING_PROBABILITY_THRESHOLD=0.1 swift sft --enable_dft_loss true ...

# Use original DFT
swift sft --enable_dft_loss true ...
```

#### References
- ProFit Paper: https://arxiv.org/pdf/2601.09195
- DFT Paper: https://arxiv.org/abs/2508.05629

### CODE REVIEW

File: swift/megatron/trainers/trainer.py [hunk 1]
```diff
@@ -59,7 +60,17 @@ def loss_func(self,
...
+            # https://arxiv.org/pdf/2601.09195
+            if hard_gating_probability_threshold := os.getenv("HARD_GATING_PROBABILITY_THRESHOLD"):
+                try:
+                    mask = (target_probs > float(hard_gating_probability_threshold)).float()
+                    target_probs = mask * loss_mask.float()
+                except ValueError:
+                    logger.warning_once('The configured environment variable HARD_GATING_PROBABILITY_THRESHOLD cannot '
+                                        'be converted to a float, so training will proceed using the default DFT mode.')
+
+            losses = losses * target_probs
```
[the remark this answers is not in the document]
Comment (maybefunctionname) [reply swift/megatron/trainers/trainer.py]: Revisions have been made according to the proposed suggestions

File: swift/trainers/utils.py [hunk 2]
```diff
@@ -85,6 +85,14 @@ def per_token_loss_func_sp(outputs, labels, enable_dft_loss=False, **kwargs) ->
...
             target_probs = torch.exp(-loss)
+            # https://arxiv.org/pdf/2601.09195
+            if hard_gating_probability_threshold := os.getenv("HARD_GATING_PROBABILITY_THRESHOLD"):
+                try:
+                    mask = (target_probs > float(hard_gating_probability_threshold)).float()
+                    target_probs = mask * (labels != -100).float()
+                except ValueError:
+                    logger.warning_once('The configured environment variable HARD_GATING_PROBABILITY_THRESHOLD cannot '
+                                        'be converted to a float, so training will proceed using the default DFT mode.')
         loss *= target_probs
```
[the remark this answers is not in the document]
Comment (maybefunctionname) [reply swift/trainers/utils.py]: Revisions have been made according to the proposed suggestions

File: swift/trainers/utils.py [hunk 3]
```diff
@@ -111,6 +119,14 @@ def per_token_loss_func(outputs, labels, enable_dft_loss: bool = False, **kwargs
...
```
[the remark this answers is not in the document]
Comment (maybefunctionname) [reply swift/trainers/utils.py]: Revisions have been made according to the proposed suggestions
